nemo/collections/nlp/models/dialogue_state_tracking/sgdqa_model.py

```diff
@@ -0,0 +1,575 @@
+        noncat_slot_status_dist = torch.nn.Softmax(dim=-1)(logit_noncat_slot_status)
+
+        noncat_slot_status = torch.argmax(logit_noncat_slot_status, axis=-1)
+        noncat_slot_status_p = torch.max(noncat_slot_status_dist, axis=-1)[0]
+
+        softmax = torch.nn.Softmax(dim=1)
+
+        scores = softmax(logit_spans)
+        start_scores, end_scores = torch.unbind(scores, dim=-1)
+
+        batch_size, max_num_tokens = end_scores.size()
+        # Find the span with the maximum sum of scores for start and end indices.
+        total_scores = torch.unsqueeze(start_scores, axis=2) + torch.unsqueeze(end_scores, axis=1)
+        # Mask out scores where start_index > end_index.
+        # device = total_scores.get_device()
+        start_idx = torch.arange(max_num_tokens, device=total_scores.get_device()).view(1, -1, 1)
+        end_idx = torch.arange(max_num_tokens, device=total_scores.get_device()).view(1, 1, -1)
+        invalid_index_mask = (start_idx > end_idx).repeat(batch_size, 1, 1)
+        total_scores = torch.where(
+            invalid_index_mask,
+            torch.zeros(total_scores.size(), device=total_scores.get_device(), dtype=total_scores.dtype),
+            total_scores,
+        )
+        max_span_index = torch.argmax(total_scores.view(-1, max_num_tokens ** 2), axis=-1)
+        max_span_p = torch.max(total_scores.view(-1, max_num_tokens ** 2), axis=-1)[0]
+
+        span_start_index = torch.floor_divide(max_span_index, max_num_tokens)
+        span_end_index = torch.fmod(max_span_index, max_num_tokens)
+
+        tensors = {
+            'example_id_num': example_id_num,
+            'service_id': service_id,
+            'intent_status': intent_status,
+            'req_slot_status': req_slot_status,
+            'cat_slot_status': cat_slot_status,
+            'cat_slot_status_p': cat_slot_status_p,
+            'cat_slot_value_status': cat_slot_value_status,
+            'noncat_slot_status': noncat_slot_status,
+            'noncat_slot_status_p': noncat_slot_status_p,
+            'noncat_slot_p': max_span_p,
+            'noncat_slot_start': span_start_index,
+            'noncat_slot_end': span_end_index,
+            'noncat_alignment_start': start_char_idx,
+            'noncat_alignment_end': end_char_idx,
+        }
+        return loss, tensors
+
+    def multi_validation_epoch_end(self, outputs: List[dict], dataloader_idx: int = 0):
+        """
+        Called at the end of validation to post process outputs into human readable format
+        Args:
+            outputs: list of individual outputs of each validation step
+            dataloader_idx: dataloader index
+        """
+        avg_loss = torch.stack([x[f'val_loss'] for x in outputs]).mean()
+        split = self._validation_names[dataloader_idx][:-1]
+        dataloader = self._validation_dl[dataloader_idx]
+        metrics = self.multi_test_epoch_end_helper(outputs=outputs, split=split, dataloader=dataloader)
```

Comment:
As multi_test_epoch_end_helper is used for validation too, how about change the name to calc_eval_metrics or something more specific?

Reply:
change to *_eval_epoch_end_helper and *eval_step_helper